lib/StaticNarrative/narrative_ref.py

```diff
@@ -32,7 +36,18 @@ def __init__(self: "NarrativeRef", ref: dict[str, str | int]) -> None:
 
         ver is not required
         """
+        # initialise these attributes to prevent errors showing up all over the codebase
+        self.wsid: int
+        self.objid: int
+        self.ver: int
+
         part_name = {"wsid": "workspace ID", "objid": "object ID", "ver": "version"}
+        # ensure all parts are available
+        missing_parts = [part for part in part_name if part not in ref]
+        if missing_parts:
+            msg = "Missing keys required to create a NarrativeRef: " + ", ".join(missing_parts)
```

Comment:
ensure that it isn't possible to do `NarrativeRef({"wsid": 12345})` and get a valid NarrativeRef out